sql/src/main/java/io/crate/execution/engine/indexing/UpsertResults.java

```diff
@@ -0,0 +1,142 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.indexing;
+
+import com.google.common.collect.ImmutableMap;
+import io.crate.data.Row;
+import io.crate.data.RowN;
+import org.apache.lucene.util.BytesRef;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+class UpsertResults {
+
+    private final Map<BytesRef, Result> resultsByUri = new HashMap<>(1);
+    private final ImmutableMap<String, BytesRef> nodeInfo;
+
+    UpsertResults() {
+        this.nodeInfo = null;
+    }
+
+    UpsertResults(ImmutableMap<String, BytesRef> nodeInfo) {
+        this.nodeInfo = nodeInfo;
+    }
+
+    void addResult(long successRowCount) {
+        Result result = getResultSafe(null);
+        result.successRowCount += successRowCount;
+    }
+
+    void addResult(BytesRef uri, @Nullable String failureMessage) {
+        Result result = getResultSafe(uri);
+        if (failureMessage == null) {
+            result.successRowCount += 1;
+        } else {
+            result.errorRowCount += 1;
+            result.updateErrorCount(failureMessage, 1L);
+        }
+    }
+
+    void addUriFailure(BytesRef uri, String uriFailure) {
+        assert uri != null : "expecting URI argument not to be null";
+        Result result = getResultSafe(uri);
+        result.sourceUriFailure = true;
+        result.updateErrorCount(uriFailure, 1L);
+    }
+
+    private Result getResultSafe(@Nullable BytesRef uri) {
+        Result result = resultsByUri.get(uri);
+        if (result == null) {
+            result = new Result();
+            resultsByUri.put(uri, result);
+        }
+        return result;
+    }
+
+    long getSuccessRowCountForNoUri() {
+        return getResultSafe(null).successRowCount;
+    }
+
+    void merge(UpsertResults other) {
+        for (Map.Entry<BytesRef, Result> entry : other.resultsByUri.entrySet()) {
+            Result result = resultsByUri.get(entry.getKey());
+            if (result == null) {
+                resultsByUri.put(entry.getKey(), entry.getValue());
+            } else {
+                result.merge(entry.getValue());
+            }
+
+        }
+    }
+
+    Iterable<Row> rowsIterable() {
+        Stream<Row> s = resultsByUri.entrySet().stream()
+            .sorted(Comparator.comparing(Map.Entry::getKey))
+            .map(e -> {
+                Result r = e.getValue();
+                if (r.sourceUriFailure) {
+                    return new RowN(new Object[]{nodeInfo, e.getKey(), null, null, r.errors});
+                } else {
+                    return new RowN(new Object[]{nodeInfo, e.getKey(), r.successRowCount, r.errorRowCount, r.errors});
+                }
+            });
+        return s::iterator;
+    }
+
+    static class Result {
+
+        private static final String ERROR_COUNT_KEY = "count";
+
+        long successRowCount = 0;
```

Comment:
all fields could be `private` as they are only accessed within class and outer class